simulation/do_nothing_plugin/src/plugin.cpp

```diff
@@ -13,15 +13,131 @@
 // limitations under the License.
 
 #include <do_nothing_plugin/plugin.hpp>
+#include <geometry/linear_algebra.hpp>
+#include <geometry/vector3/hypot.hpp>
 
 namespace entity_behavior
 {
 void DoNothingBehavior::configure(const rclcpp::Logger &) {}
 
-void DoNothingBehavior::update(double current_time, double)
+void DoNothingBehavior::update(double current_time, double step_time)
 {
+  setCurrentTime(current_time);
+  setStepTime(step_time);
   entity_status_->setTime(current_time);
-  setUpdatedStatus(entity_status_);
+  if (getRequest() == traffic_simulator::behavior::Request::FOLLOW_POLYLINE_TRAJECTORY) {
+    followPolylineTrajectory();
+  } else {
+    setUpdatedStatus(entity_status_);
+  }
+}
+
+void DoNothingBehavior::checkPolylineTrajectory()
+{
+  if (const auto trajectory = getPolylineTrajectory()) {
+    if (trajectory->closed) {
+      THROW_SIMULATION_ERROR("Currently, closed trajectory does not supported.");
+    }
+    if (!trajectory->dynamic_constraints_ignorable) {
+      THROW_SIMULATION_ERROR(
+        "Currently, dynamic_constraints_ignorable = true (in OpenSCENARIO, followingMode = "
+        "follow) does not support in DoNothingBehavior.");
+    }
+    if (std::abs(trajectory->initial_distance_offset) > std::numeric_limits<double>::epsilon()) {
+      THROW_SIMULATION_ERROR(
+        "Currently, initial_distance_offset should be 0 when following trajectory in "
+        "DoNothingBehavior.",
+        "You specified : ", trajectory->initial_distance_offset);
+    }
+    if (trajectory->shape.vertices.empty()) {
+      THROW_SIMULATION_ERROR(
+        "FollowPolylineTrajectory is requested, but trajectory points are empty. Please check "
+        "description of the scenario.");
+    }
+  } else {
+    THROW_SIMULATION_ERROR(
+      "Traffic simulator send requests of FollowTrajectory, but the trajectory is empty.",
+      "This message is not originally intended to be displayed, if you see it, please "
+      "contact the developer of traffic_simulator.");
+  }
+}
+
+void DoNothingBehavior::followPolylineTrajectory()
+{
+  updated_status_ = std::shared_ptr<traffic_simulator::CanonicalizedEntityStatus>();
+  checkPolylineTrajectory();
+
+  if (const auto trajectory = getPolylineTrajectory()) {
+    for (size_t i = 0; i < trajectory->shape.vertices.size() - 1; i++) {
```

Comment:
The algorithm "iterate over vertices by index and break when it finds one that matches the condition" seems very bug-prone. This block should be written using std::adjacent_find.

Reply:
Fixed in https://github.com/tier4/scenario_simulator_v2/pull/1217/commits/fee36c909165f967703b5eace2f36f3e7070b47c